libyul/optimiser/StackToMemoryMover.cpp

```diff
@@ -0,0 +1,204 @@
+/*
+	This file is part of solidity.
+
+	solidity is free software: you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation, either version 3 of the License, or
+	(at your option) any later version.
+
+	solidity is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY; without even the implied warranty of
+	MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+#include <libyul/optimiser/StackToMemoryMover.h>
+#include <libyul/optimiser/NameDispenser.h>
+#include <libyul/backends/evm/EVMDialect.h>
+
+#include <libyul/AsmData.h>
+
+#include <libsolutil/CommonData.h>
+#include <libsolutil/Visitor.h>
+
+
+using namespace std;
+using namespace solidity;
+using namespace solidity::yul;
+
+namespace
+{
+void appendMemoryStore(
+	vector<Statement>& _statements,
+	langutil::SourceLocation const& _loc,
+	YulString _mpos,
+	Expression _value
+)
+{
+	_statements.emplace_back(ExpressionStatement{_loc, FunctionCall{
+		_loc,
+		Identifier{_loc, "mstore"_yulstring},
+		{
+			Literal{_loc, LiteralKind::Number, _mpos, {}},
+			std::move(_value)
+		}
+	}});
+}
+}
+
+StackToMemoryMover::StackToMemoryMover(
+	OptimiserStepContext& _context,
+	u256 _reservedMemory,
+	map<YulString, map<YulString, uint64_t>> const& _memorySlots
+): m_reservedMemory(std::move(_reservedMemory)), m_memorySlots(_memorySlots), m_nameDispenser(_context.dispenser)
+{
+	auto const* evmDialect = dynamic_cast<EVMDialect const*>(&_context.dialect);
+	yulAssert(
+		evmDialect && evmDialect->providesObjectAccess(),
+		"StackToMemoryMover can only be run on objects using the EVMDialect with object access."
+	);
+
+	if (m_memorySlots.count(YulString{}))
+		m_currentFunctionMemorySlots = &m_memorySlots.at(YulString{});
+}
+
+void StackToMemoryMover::operator()(FunctionDefinition& _functionDefinition)
+{
+	map<YulString, uint64_t> const* saved = m_currentFunctionMemorySlots;
+	if (m_memorySlots.count(_functionDefinition.name))
+	{
+		m_currentFunctionMemorySlots = &m_memorySlots.at(_functionDefinition.name);
+		for (TypedName const& param: _functionDefinition.parameters + _functionDefinition.returnVariables)
+			if (m_currentFunctionMemorySlots->count(param.name))
+			{
+				// TODO: we cannot handle function parameters yet.
+				m_currentFunctionMemorySlots = nullptr;
+				break;
+			}
+	}
+	else
+		m_currentFunctionMemorySlots = nullptr;
```

Comment:
So the global scope doesn't matter at all when inside a function?

Reply:
In Yul you don't have access to global variables from inside functions: so yes, the global scope doesn't matter inside function bodies.